file/file_prefetch_buffer.cc

```diff
@@ -166,14 +206,77 @@ bool FilePrefetchBuffer::TryReadFromCache(const IOOptions& opts,
 #endif
         return false;
       }
-      readahead_size_ = std::min(max_readahead_size_, readahead_size_ * 2);
+      prefetched = true;
     } else {
       return false;
     }
   }
-  UpdateReadPattern(offset, n);
+  UpdateReadPattern(offset, n, false /*decrease_readaheadsize*/);
+
+  // Prefetch asynchronously if
+  // - Prefetch is called for current data block and,
+  // - is implicit_auto_readahead_
+  // - is_adaptive_readahead_ is set true and,
+  // - After synchronous Prefetch, result.size() == read_len indicating not end
+  // of file.
+  if (prefetched && implicit_auto_readahead_ && is_adaptive_readahead_ &&
```

Comment:
IIUC, async Prefetch will be called only if we had to prefetch some data in this call to TryReadFromCache. This means that the pattern will be a blocking read for n+readahead_size_ and an async read for another readahead_size_/2. I don't know if it'll be that beneficial, since the blocking read will incur latency. Its not much different from doing a single blocking read for n+readahead_size_+readahead_size_/2.

Can we do something like the following instead? Assume that there is readahead_size_ bytes in the prefetch buffer. As soon as readahead_size_/2 bytes is consumed from the buffer, start an async prefetch for the range buffer_offset_+readahead_size_/2 to buffer_offset_+readahead_size_. IOW, we don't wait to start an async prefetch until we have to do a blocking prefetch. We can start the next async prefetch while the iterator is still consuming data that's already prefetched. This way, we can have true overlap of compute and IO.

Reply:
Also, I think its better to use 2 AlignedBuffers, in order to avoid too much copying of data due to RefitTail. TryReadFromCache can read from one buffer while the async prefetch is filling the other buffer. It also simplifies the synchronization problem, since the 2nd buffer can be updated by the async read callback while the main thread is consuming data from the 1st buffer.

Reply:
> Also, I think its better to use 2 AlignedBuffers, in order to avoid too much copying of data due to RefitTail. TryReadFromCache can read from one buffer while the async prefetch is filling the other buffer. It also simplifies the synchronization problem, since the 2nd buffer can be updated by the async read callback while the main thread is consuming data from the 1st buffer.

Yes, I will use 2 AlignedBuffers. I will make things much more simple and easy to handle.